api/app/controllers/transaction.controller.js

```diff
@@ -0,0 +1,137 @@
+exports.create = async (req, res) => {
+  // check if any fields are empty
+  const requiredFields = ["userId", "amount", "transactionType"];
+  if (
+    req.body.transactionType &&
+    req.body.transactionType === TRANSACTION_TYPE.CREDIT
+  ) {
+    requiredFields.push("reviewId");
+  }
+  const missingFields = requiredFields.filter((field) => !req.body[field]);
+  if (missingFields.length) {
+    res.status(400).send({
+      message: `Missing required fields: ${missingFields.join(", ")}`,
+    });
+    return;
+  }
+
+  const currentTime = new Date();
+
+  // check if review exists
+  if (req.body.reviewId) {
+    const review = await Review.findOne({
+      where: {
+        id: req.body.reviewId,
+      },
+    });
+    if (!review) {
+      res.status(404).send({
+        message: `Could not create transaction: review with id=${req.body.reviewId} does not exist`,
+      });
+      return;
+    }
+  }
+
+  // check if user wallet exists
+  const userWallet = await Wallet.findOne({
+    where: {
+      userId: req.body.userId,
+    },
+  });
+  if (!userWallet) {
+    res.status(404).send({
+      message: `Could not create transaction: wallet with id=${req.body.walletId} does not exist`,
+    });
+    return;
+  }
+  const transactionId = generateTransactionId();
+  const transaction = {
+    id: transactionId,
+    reviewId: req.body.reviewId ?? null,
+    amount: +req.body.amount,
+    transactionType: req.body.transactionType,
+    transactionStatus: TRANSACTION_STATUS.SUCCESS,
+    walletId: userWallet.id,
+    timestamp: currentTime,
+  };
+
+  const newWalletBalance =
+    transaction.transactionType === TRANSACTION_TYPE.CREDIT
+      ? userWallet.balance + transaction.amount
+      : userWallet.balance - transaction.amount;
+
+  try {
+    const transactionResult = await Transaction.create(transaction);
+    await Wallet.update(
+      { balance: newWalletBalance, updatedAt: currentTime },
+      { where: { id: userWallet.id } }
+    );
+    return res.send(transactionResult);
+  } catch (error) {
+    transaction.transactionStatus = TRANSACTION_STATUS.FAILURE;
+    Transaction.create(transaction);
+    res.status(500).send({
+      message:
+        err.message || "Some error occurred while creating the Transaction.",
+    });
+  }
+};
+
+exports.findAll = async (req, res) => {
+  const { userId, status, type } = req.query;
+  // validate fields
+  if (!userId) {
+    res.status(400).send({
+      message: `Missing required fields: userId`,
+    });
+    return;
+  } else if (isNaN(userId)) {
+    res.status(400).send({
+      message: `userId is of type ${typeof userId}. userId should be a number.`,
+    });
+  }
+
+  const userWallet = await Wallet.findOne({ where: { userId: userId } });
+  if (!userWallet) {
+    res.status(404).send({
+      message: `Transaction for userId=${userId} does not exist`,
+    });
+    return;
+  }
+
+  let condition = { walletId: userWallet.id };
+  if (status && type) {
```

Comment:
query params reduces all types to string, i.e extracheck for undefined or null as string. e.g if status is passed as undefined from where it is called, it is received as `status: "undefined"`

Reply:
I don't think we need an extra check. if status is undefined or null, this condition to filter based on status won't be appended and we don't get an error. Am I missing something?

Reply:
It(query params) is passed as strings so a boolean will check `Boolean("undefined")` instead of `Boolean(undefined)` and the former one will be true so it will append the condition. You can try it out, I ran into the issue when I was writing status query on reviews